File: lib/perfmonger/command/play.rb

require 'optparse'
require 'json'

module PerfMonger
module Command

class PlayCommand < BaseCommand
  register_command 'play', "Play a perfmonger log file in JSON"

  def initialize
    @parser = OptionParser.new
    @parser.banner = <<EOS
Usage: perfmonger play [options] LOG_FILE

Options:
EOS

    @color = false
    @parser.on("-c", "--color", "Use colored JSON ouptut") do
      @color = true
    end

    @pretty = false
    @parser.on("-p", "--pretty", "Use human readable JSON ouptut") do
      @pretty = true
    end

    @disk_only_regex = nil
    @parser.on('--disk-only REGEX', "Select disk devices that matches REGEX (Ex. 'sd[b-d]')") do |regex|
      @disk_only_regex = regex
    end
  end

  def parse_args(argv)
    @parser.parse!(argv)

    if argv.size == 0
      puts("ERROR: PerfMonger log file is required")
      puts(@parser.help)
      exit(false)
    end

    @logfile = argv.shift
    if ! File.exists?(@logfile)
      puts("ERROR: No such file: #{@logfile}")
      puts(@parser.help)
      exit(false)
    end
  end

  def run(argv)
    parse_args(argv)

    @player_bin = ::PerfMonger::Command::CoreFinder.player()

    if ! @player_bin
      puts("[ERROR] no executable binary found.")
      exit(false)
    end

    cmd = [@player_bin]
    if @color
      cmd << "-color"
    end
    if @pretty
      cmd << "-pretty"
    end

    if @disk_only_regex
      cmd << "-disk-only"
      cmd << @disk_only_regex
    end

    cmd << @logfile

    Process.exec(*cmd)
  end
end

end # module Command
end # module PerfMonger

File: core/cmd/perfmonger/play.go
package main

import (
	"fmt"
	"os"

	"github.com/spf13/cobra"
	"github.com/hayamiz/perfmonger/core/cmd/perfmonger-core/player"
)

// playCommand represents the play command with direct PlayerOption setting
type playCommand struct {
	// Direct field (no embedding) for maximum efficiency
	PlayerOpt *player.PlayerOption
	
	// No Ruby-specific options for play command
}

// newPlayCommandStruct creates playCommand with Ruby-compatible defaults
func newPlayCommandStruct() *playCommand {
	return &playCommand{
		PlayerOpt: player.NewPlayerOption(),
	}
}

// validateAndSetLogfile validates the logfile argument using cobra's PreRunE approach
func (cmd *playCommand) validateAndSetLogfile(args []string) error {
	if len(args) == 0 {
		// No file argument: read from stdin (default Logfile is "-")
		return nil
	}

	// Take the first argument as log file
	cmd.PlayerOpt.Logfile = args[0]

	// Check if file exists
	if _, err := os.Stat(cmd.PlayerOpt.Logfile); os.IsNotExist(err) {
		return fmt.Errorf("no such file: %s", cmd.PlayerOpt.Logfile)
	}

	return nil
}

// run executes the play command with direct API calls
func (cmd *playCommand) run() error {
	if os.Getenv("PERFMONGER_DEBUG") != "" {
		fmt.Fprintf(os.Stderr, "[debug] running player with options: %+v\n", cmd.PlayerOpt)
	}
	
	// Direct API call - no conversion needed
	player.RunWithOption(cmd.PlayerOpt)
	return nil
}


// newPlayCommand creates the play subcommand with direct cobra setting
func newPlayCommand() *cobra.Command {
	playCmd := newPlayCommandStruct()
	
	cmd := &cobra.Command{
		Use:   "play [options] LOG_FILE",
		Short: "Play a recorded perfmonger session",
		Long:  `Play a perfmonger log file in JSON`,
		PreRunE: func(cmd *cobra.Command, args []string) error {
			// Validation moved to PreRunE for cobra integration
			return playCmd.validateAndSetLogfile(args)
		},
		RunE: func(cmd *cobra.Command, args []string) error {
			// Direct execution - no additional validation needed
			return playCmd.run()
		},
	}
	
	// Direct cobra flag setting to PlayerOption fields (no conversion needed)
	cmd.Flags().BoolVarP(&playCmd.PlayerOpt.Color, "color", "c", playCmd.PlayerOpt.Color,
		"Use colored JSON output")
	cmd.Flags().BoolVarP(&playCmd.PlayerOpt.Pretty, "pretty", "p", playCmd.PlayerOpt.Pretty,
		"Use human readable JSON output")
	cmd.Flags().StringVar(&playCmd.PlayerOpt.DiskOnly, "disk-only", playCmd.PlayerOpt.DiskOnly,
		"Select disk devices that matches REGEX (Ex. 'sd[b-d]')")
	
	cmd.SetUsageTemplate(subCommandUsageTemplate)
	return cmd
}